include/swift/Syntax/SyntaxData.h

```diff
@@ -76,19 +76,30 @@ class SyntaxData {
   const AbsoluteRawSyntax AbsoluteRaw;
   RC<RefCountedBox<SyntaxData>> Parent;
 
+  /// If this node is the root of a Syntax tree (i.e. \c Parent is \c nullptr ),
+  /// the arena in which this node's \c RawSyntax node has been allocated.
+  /// This keeps this \c RawSyntax nodes referenced by this tree alive.
+  RC<SyntaxArena> Arena;
```

Comment:
Not in this PR: since `Parent` and `Arena` are mutually exclusive, it would be great if these can be a tagged union.

Reply:
There’s a reason why it’s not a union. In the next PR, I’ll split `SyntaxData` into a `SyntaxDataRef` that doesn’t retain the arena (but has a parent) and a subclass `SyntaxData` which retains the arena. So I can’t easily reuse the storage.